File: src/lizi2d/interp.py
from __future__ import annotations

import numpy as np

from .grid import Grid2D
from .particles import ParticleState


def gather_field_to_particles_bilinear(
    grid: Grid2D,
    particles: ParticleState,
    Ex: np.ndarray,
    Ey: np.ndarray,
) -> tuple[np.ndarray, np.ndarray]:
    """
    将定义在网格节点上的电场矢量场（Ex, Ey）通过双线性插值（gather）
    采样到粒子位置，得到粒子的受力分量（q=1 时 F=E）。

    前提/假设：
    - 粒子坐标 (x,y) 在周期域内（越界后自动按周期包裹）
    - Ex/Ey 的 shape 为 (nx, ny)
    - 插值权重体系与 scatter_unit_charges_to_grid 保持一致（双线性权重）

    返回：
      fx, fy：shape = (N,) 的粒子电场采样值
    """
    nx, ny = grid.shape

    # 周期性包裹粒子坐标
    lx = nx * grid.dx
    ly = ny * grid.dy
    xw = np.mod(particles.x, lx)
    yw = np.mod(particles.y, ly)
    gx = xw / grid.dx
    gy = yw / grid.dy

    i0 = np.floor(gx).astype(np.int64) % nx
    j0 = np.floor(gy).astype(np.int64) % ny
    i1 = (i0 + 1) % nx
    j1 = (j0 + 1) % ny

    fx = gx - np.floor(gx)
    fy = gy - np.floor(gy)

    wx0 = 1.0 - fx
    wx1 = fx
    wy0 = 1.0 - fy
    wy1 = fy

    # 双线性插值
    Ex_p = (
        Ex[i0, j0] * wx0 * wy0
        + Ex[i1, j0] * wx1 * wy0
        + Ex[i0, j1] * wx0 * wy1
        + Ex[i1, j1] * wx1 * wy1
    )
    Ey_p = (
        Ey[i0, j0] * wx0 * wy0
        + Ey[i1, j0] * wx1 * wy0
        + Ey[i0, j1] * wx0 * wy1
        + Ey[i1, j1] * wx1 * wy1
    )
    return Ex_p, Ey_p

File: src/core/interp.rs
use ndarray::Array1;
use ndarray::Array2;

use crate::core::grid::Grid2D;
use crate::core::particles::ParticleState;

/// 将定义在网格节点上的电场矢量场 (Ex, Ey) 通过双线性插值采样到粒子位置
///
/// 返回: (fx, fy) — shape 均为 (N,) 的粒子受力分量
pub fn gather_field_to_particles_bilinear(
    grid: &Grid2D,
    particles: &ParticleState,
    ex: &Array2<f64>,
    ey: &Array2<f64>,
) -> (Array1<f64>, Array1<f64>) {
    let (nx, ny) = grid.shape();
    let n = particles.len();
    let lx = grid.lx();
    let ly = grid.ly();

    let mut fx = Array1::zeros(n);
    let mut fy = Array1::zeros(n);

    for p in 0..n {
        // 周期包裹
        let xw = ((particles.x[p] % lx) + lx) % lx;
        let yw = ((particles.y[p] % ly) + ly) % ly;
        let gx = xw / grid.dx;
        let gy = yw / grid.dy;

        let i0 = (gx.floor() as i64).rem_euclid(nx as i64) as usize;
        let j0 = (gy.floor() as i64).rem_euclid(ny as i64) as usize;
        let i1 = (i0 + 1) % nx;
        let j1 = (j0 + 1) % ny;

        let ftx = gx - gx.floor();
        let fty = gy - gy.floor();

        let wx0 = 1.0 - ftx;
        let wx1 = ftx;
        let wy0 = 1.0 - fty;
        let wy1 = fty;

        // 双线性插值
        fx[p] = ex[[i0, j0]] * wx0 * wy0
            + ex[[i1, j0]] * wx1 * wy0
            + ex[[i0, j1]] * wx0 * wy1
            + ex[[i1, j1]] * wx1 * wy1;

        fy[p] = ey[[i0, j0]] * wx0 * wy0
            + ey[[i1, j0]] * wx1 * wy0
            + ey[[i0, j1]] * wx0 * wy1
            + ey[[i1, j1]] * wx1 * wy1;
    }

    (fx, fy)
}